dartagnan/src/main/java/com/dat3m/dartagnan/utils/symmetry/CoSymmetryBreaking.java

```diff
@@ -0,0 +1,244 @@
+            - We create a row of the form (exec(W_1), co(W_1, W_2), co(W_1, W_3), ...);(other co-vars independent of W_1)
+            - We encode complete row-symmetry breaking on this row and its symmetric rows
+              (obtained by the transpositions (i <-> i+1))
+        (4) Due to the order of symmetry breaking, we have "co(W_i, W_j) is a must-edge for all i < j".
+
+    NOTE: The addition of "exec(W_1)" to the front of the row allows the following reasoning:
+          - If co(W_1, W_3) happens, then so does exec(W_3). By symmetry breaking, we have to have exec(W_2) as well.
+            But then, also by symmetry breaking, we have co(W_1, W_2) and co(W_2, W_3).
+          - In other words, it allows us to assume that all threads that execute their W_i are adjacent
+            and we do not get "gaps" if some intermediate thread does not execute its write W
+ */
+public class CoSymmetryBreaking {
+
+    private static final Logger logger = LogManager.getLogger(CoSymmetryBreaking.class);
+
+    private final VerificationTask task;
+    private final ThreadSymmetry symm;
+    private final Relation co;
+
+    private final Map<EquivalenceClass<Thread>, Info> infoMap;
+
+    // Stores information per symmetry class
+    private static class Info {
+        List<Thread> threads; // Sorted list of threads of the symmetry class
+        List<Store> writes; // Writes ordered by importance / breaking order
+        boolean hasMustEdges; // True if the coherences of the first write in <writes> can be broken via must-edges.
+    }
+
+    public CoSymmetryBreaking(VerificationTask task) {
+        this.task = task;
+        this.symm = task.getThreadSymmetry();
+        this.co = task.getMemoryModel().getRelationRepository().getRelation(RelationNameRepository.CO);
+        infoMap = new HashMap<>();
+    }
+
+    // Computes data for symmetry breaking and must-set generation
+    // Returns TRUE, if must-sets can be generated
+    public void initialize() {
+        infoMap.clear();
+        for (EquivalenceClass<Thread> symmClass : symm.getNonTrivialClasses()) {
+            initialize(symmClass);
+        }
+    }
+
+    private void initialize(EquivalenceClass<Thread> symmClass) {
+        Info info = new Info();
+        infoMap.put(symmClass, info);
+
+        // Get symmetric threads and order them by tid
+        List<Thread> symmThreads = new ArrayList<>(symmClass);
+        symmThreads.sort(Comparator.comparingInt(Thread::getId));
+        info.threads = symmThreads;
+
+        // Get stores of first thread
+        List<Store> writes = symmThreads.get(0).getCache().getEvents(FilterBasic.get(EType.WRITE))
+                .stream().map(Store.class::cast).collect(Collectors.toList());
+
+        // Symmetric writes that cannot alias are related to e.g. thread-creation
+        // We do not want to break on those
+        writes.removeIf(w -> !Store.canAddressTheSameLocation(w, (Store)symm.map(w, symmThreads.get(1))));
```

Comment:
Yes, that is true. While there is the theoretical case that the first two thread could alias but the others do not, it is a rather rare occurence because all threads are symmetric. 
Basically, you expect that if they alias then it is because the writes are independent of any input. The case where this reasoning breaks is if the writes are dependent on the inputs but the first two threads get the same input while the remaining do not (which is weird). In either case, even if we encode a breaking on those coherences, it is not wrong and will just increase the encoding size (that breaking will never occur if they cannot alias and it will just be a "dead clause" in the encoding)